src/main/java/org/araymond/joal/core/ttorrent/client/ConnectionHandler.java

```diff
@@ -25,71 +34,96 @@ public class ConnectionHandler {
     private static final String[] IP_PROVIDERS = new String[]{
+            "http://whatismyip.akamai.com",
+            "http://ipecho.net/plain",
             "http://ip.tyk.nu/",
             "http://l2.io/ip",
             "http://ident.me/",
-            "http://icanhazip.com/"
+            "http://icanhazip.com/",
+            "https://api.ipify.org",
+            "https://ipinfo.io/ip",
+            "https://checkip.amazonaws.com"
     };
 
     public int getPort() {
-        return this.channel.socket().getLocalPort();
+        return this.listenChannel.socket().getLocalPort();
     }
 
     public void start() throws IOException {
-        this.channel = this.bindToPort();
-        final int port = this.channel.socket().getLocalPort();
-        log.info("Listening for incoming peer connections on port {}.", port);
+        this.listenChannel = this.bindToPort();
+        log.info("Listening for incoming peer connections on port {}", getPort());
 
         this.ipAddress = fetchIp();
-        log.info("Ip reported to tracker will be: {}", this.getIpAddress().getHostAddress());
+        log.info("IP reported to tracker will be: {}", this.getIpAddress().getHostAddress());
 
+        // TODO: use @Scheduled
         this.ipFetcherThread = new Thread(() -> {
-            while (this.ipFetcherThread == null || !this.ipFetcherThread.isInterrupted()) {
+            while (this.ipFetcherThread != null && !this.ipFetcherThread.isInterrupted()) {
                 try {
-                    // Sleep for one hour and a half.
-                    Thread.sleep(1000 * 5400);
+                    MINUTES.sleep(90);  // TODO: move to config
                     this.ipAddress = this.fetchIp();
                 } catch (final UnknownHostException e) {
-                    log.warn("Faield to fetch Ip", e);
+                    log.warn("Failed to fetch external IP", e);
                 } catch (final InterruptedException e) {
-                    log.info("Ip fetcher thread has been stopped.");
+                    log.info("IP fetcher thread has been stopped");
                 }
             }
         });
 
         this.ipFetcherThread.start();
+
+        this.serverThread = new Thread(() -> {
+            try {
+                Selector selector = Selector.open();
+                this.listenChannel.register(selector, SelectionKey.OP_ACCEPT);
+
+                while (this.listenChannel != null && this.serverThread != null && !this.serverThread.isInterrupted()) {
+                    selector.select();
+                    for (SelectionKey key : selector.selectedKeys()) {
+                        if (key.isAcceptable()) {
+                            this.listenChannel.accept();
+                        }
+                    }
+                }
+            } catch (IOException ioe) {
+                log.debug("Problem processing incoming tracker connection", ioe);
+                throw new RuntimeException(ioe);
```

Comment:
Suspect throwing here is rather pointless... would expect IOEx to be thrown maybe also when the client closes the connection.